### PR TITLE
[RHMAP-5128] Change grunt-fh-build to use ESLint instead of JSHint

### PR DESCRIPTION
**Pull request by paolobueno**
Also removed `plato` since AFAIK `eslint` has a builtin rule for complexity.

I have a question: why do we target the `lib/` and `bin/` directories specifically? on `fh-fhc` for instance we have a lot of small rule violations inside `test/`, for instance. I tried changing the target configuration to `['**/*.js']` but the task broke for some reason.

I've also tried creating another variation of the task like `grunt eslint:fix`, but I think it would always run alongside the main task, and we should probably run `eslint --fix` straight from the cli anyways.

I've bumped the minor in this PR, since it's kind of a breaking change. I'll also squash the commits before merging.

ping @jasonmadigan, @wei-lee, @wtrocki, @ziccardi, @odra

### CODE REVIEW

**Comment by fh-automatron** [comment]
### Pull Request Quality Guidelines

As we continue to focus on quality, here are a few things to keep in mind:
- A component owner should review your change (we've tagged these above)
- This change should be deployed to a grid somewhere (via `fhcap-cli`) - please provide a link to this
- Automated testing has been added or updated for this change (e.g. UART, Unit/Acceptance tests)
- Ensure JIRA is up to date (e.g. deployment notes, tester notes, fix version, release notes)
- There should be an associated fhcap PR with this change - please provide a link to this
- Any SonarQube issues should be addressed where applicable (use your own judgement)

Please ping a component owner (see `fh contrib`) when your builds are complete, changes are deployed and you're ready for a review.

**Comment by wtrocki** [comment]
Thanks for adding jslint support. Just one comment from me: I do not think that we should remove jshint support, at least for now. 

#### Reasons why we should not remove jshit:
- This module is used by various components from multiple teams, organizations etc. 
- Removing jshit will just break builds for all components that aren't integrated with eslint - 
  still lots of them. This may be fixed by using default everywhere, but it's still not possible in all cases.
- We will lose ability to build older versions that base on jshint (when specifying direct target)

#### Proposed solution:

Add framework detection - basing on config we will know what framework is used and determine
if we should use jshint or eslint. Solution should default to jshit to support backwards compatibility.
New option and new framework should be documented.

#### Note

This PR will require changes in job builders first. This is the reason why PR build is failing :)

File: tasks/fh-build.js [hunk 1]
```diff
@@ -100,26 +101,10 @@ module.exports = function(grunt) {
...
-      fh: {
-        options : {
-          jshint : grunt.file.readJSON('.jshintrc')
-        },
-        files: {
-          'plato': ['lib/**/*.js']
-        }
-      }
+    eslint: {
+      target: lintTargets
```
Comment (grdryn) [inline tasks/fh-build.js:107]: Do you think there would be benefit to having this 'lintTargets' be something that can be overridden at a per-project basis? I imagine it's possible, but I haven't looked into it.

File: tasks/fh-build.js [hunk 2]
```diff
@@ -403,12 +387,10 @@ module.exports = function(grunt) {
         'shell:fh-run-array:integrate_cover', 'shell:fh-run-array:accept_cover',
         'shell:fh-report-cov:lcov', 'shell:fh-report-cov:cobertura'
       ]);
-    } else if (this.target === 'analysis') {
-      grunt.task.run(['plato:fh']);
```
Comment (wtrocki) [inline tasks/fh-build.js:407]: This is going to break some jenkins builds.

**Comment by wei-lee** [comment]
@paolobueno This change looks good. But as @wtrocki said, it will be better if we can support both, and automatically choose the right linting tool based on configuration.

If that proves to be problematic, we can just use eslint, but we need to bump the major version to make people aware that it is not backward-compatible. It they choose to use the latest version, they will need to update the component to use eslint. Maybe this will work better, as eventually we don't want to have 2 linting tools?

@jasonmadigan @grdryn WDYT?

**Comment by grdryn** [comment]
Like @wtrocki says, these changes will require changes in jenkins-bob-builder also, to make sure that jshint and plato steps are removed.

@wtrocki I don't think adding any framework support is necessary: I think we could update all projects in jenkins-bob-builder, then leave jshint and plato options here, rather than removing them. What do you think? Would framework support still be necessary? I think ideally we don't have two things that do the same thing for too long.

Also, would we be able to identify which projects could need builds of older versions, and of those, which depend on jshint or plato (and how strong of a dependency this is)? If an old version of something gets a hotfix, what job config will it use, the old version or the latest version? I think the latest version but I could be wrong. Either way, depending on the version matcher for grunt-fh-build in the package.json of the older module version, do we know which version of this plugin will be pulled in? (this reminds me, if this PR is removing functionality, it should bump major! :))

**Comment by wtrocki** [comment]
@grdryn Fully agree. I just wanted to be sure that builds will pass. When bumping minor version we will have full control over what framework we will use. We may change to new version only for those components that are already migrated to eslint.

**Comment by paolobueno** [comment]
Hello all! I figured this change would affect CI builds. I think they should reference more generic/umbrella tasks, so we have more freedom to change them. Like @wtrocki pointed out to me happens on [the jenkins config file](https://github.com/fheng/jenkins-bob-builder/blob/d71c8874c9bcaf64052fab491aaa28bbf9392da4/jobs/node-pullrequests/projects.yaml#L74), IMO there's no reason it should be referencing (coupled to) `plato` directly, since we have the semantic task name of `fh:analysis` that could be changed to use another equivalent tool anytime.

So what I was wondering is if we could have something like

``` javascript
grunt.registerTask('lint', function() {
  if (fs.existsSync('.jshintrc')) {
    grunt.log.error('jshint usage deprecated, import the base .eslintc.json from {fheng/help url}');
    return grunt.task.run('jshint');
  }
  if (fs.existsSync('.eslintrc.*')) {
    return grunt.task.run('eslint');
  }
  grunt.fail.warn('no linting config found!');
});
```

Of course this also breaks some CI builds that point straight to `jshint`, ideally we should be able to alias it while 'shadowing' the original task:

``` javascript
// delegate to `lint` but ask for specific linter, shadows/wraps task from npm module
grunt.registerTask('jshint', "lint 'jshint'");

grunt.registerTask('lint', function(linter){
  ... // use jshint task from npm module but print deprecation message
});
```

But I'm not sure this is possible in grunt.

**Comment by paolobueno** [comment]
On the current PR I also assumed `fh:analysis` and `plato` are made obsolete by eslint. They could be changed to a noop, that just prints a deprecation message but doesn't fail. i.e. doesn't require `--force`.

But now that I've taken a better look at plato I'm not sure whether we use the more advanced metrics available through [escomplex](https://github.com/jared-stilwell/escomplex/blob/master/README.md#metrics), but they're definitely more complex than the [eslint rule](https://github.com/eslint/eslint/blob/master/lib/rules/complexity.js), though this one is probably more akin to what SonarQube is doing on it's complexity rule, and the rest of the metrics are kind of covered by other rules on SonarQube.

If it has value we should keep `plato`.

**Comment by grdryn** [comment]
@paolobueno I'm not sure that plato has value that we use. I don't think anyone ever looks at the plato reports, while for now we at least look at sonarqube occassionally. @jasonmadigan @wei-lee do you guys think that plato has been valuable, or should we just rip it out?

On the topic of jshint/eslint, if they do the same thing, I'd rather we just use one. It seems like that's going to be eslint going forward? If so, I think just take out jshint. If something breaks, it's probably a trivial fix (in the worst case where it turns out that jshint is load-bearing, individual projects can just include it themselves, rather than trying to make this plugin cover all exceptional cases). What do you guys think?

**Comment by jasonmadigan** [comment]
@grdryn Yes, I think we should remove plato. It's largely superseded by ESLint + SonarQube, and nobody looks at the plato reports. They're also not graphed over time (like SonarQube), so they're pretty useless in my mind.

**Comment by grdryn** [comment]
@paolobueno I've cherry-picked what I think are the important bits from here into #13. I was making loads of changes there, then remembered that you had these changes here. I'll send a mail asking for feedback on that one.

**Comment by paolobueno** [comment]
I'm glad these changes could be of use, I'll be closing this PR :)

**Pull request closed without merging**